control-plane/csi-driver/src/context.rs

```diff
@@ -173,10 +205,114 @@ impl TryFrom<&HashMap<String, String>> for CreateParams {
             None => 1,
         };
 
+        let volume_group = Parameters::volumne_group(args.get(Parameters::VolumeGroup.as_ref()))
+            .map_err(|_| {
+                tonic::Status::invalid_argument(
+                    "Invalid `openebs.io/volumegroup` value, expected a bool",
+                )
+            })?;
+
+        let volume_group_name = if volume_group.unwrap_or(false) {
+            generate_volume_group_name(
+                args.get(Parameters::PvcName.as_ref()).cloned(),
+                args.get(Parameters::PvcNamespace.as_ref()).cloned(),
+            )
+        } else {
+            None
+        };
+
         Ok(Self {
             publish_params,
             share_protocol,
             replica_count,
+            volume_group: volume_group_name,
         })
     }
 }
+
+// Generate a volume group name from the parameters.
+// 1. Both pvc name and ns should be valid.
+// 2. Pvc name should follow the sts pvc naming convention.
+fn generate_volume_group_name(pvc_name: Option<String>, pvc_ns: Option<String>) -> Option<String> {
+    match (pvc_name.clone(), pvc_ns.clone()) {
+        (Some(pvc_name), Some(pvc_ns)) => {
+            let re = Regex::from_str(K8S_STS_PVC_NAMING_REGEX);
+            if let Ok(regex) = re {
+                if regex.is_match(pvc_name.as_str()) {
+                    if let Some(captures) = regex.captures(pvc_name.as_str()) {
+                        if let Some(common_binding) = captures.get(1) {
+                            return Some(format!("{pvc_ns}/{}", common_binding.as_str()));
+                        }
+                    }
+                }
+            }
+            warn!("PVC Name: {pvc_name} is not a valid statefulset pvc naming format, not triggering statefulset volume replica anti-affinity");
+            None
+        }
+        _ => {
+            warn!("Invalid PVC Name: {:?} or PVC Namespace: {:?}, not triggering statefulset volume replica anti-affinity", pvc_name, pvc_ns);
+            None
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::context::generate_volume_group_name;
+
+    struct TestSuite {
+        pvc_name: Option<String>,
+        pvc_namespace: Option<String>,
+        result: Option<String>,
+    }
+
+    impl TestSuite {
+        fn new(
+            pvc_name: Option<String>,
```

Comment:
why not take Option<&str> to avoid all the .to_string() below?

Reply:
Had tried with &str, that's causing a issues with lifetime.

Reply:
On the new you'd need to map to `String`.

Reply:
Ohh, right. Done.